### ISSUE TITLE
[DRAFT ][Bounty $1500] Diffusion Drive bring up using TTNN APIs

### ISSUE THREAD

**Issue by dvartaniansTT** [tenstorrent/tt-metal#31269]
### 📝 Background
This bounty is for bringing up [DiffusionDrive](https://github.com/hustvl/DiffusionDrive?utm_source=chatgpt.com)
 using TTNN APIs on Tenstorrent hardware (Wormhole or Blackhole).
DiffusionDrive is a truncated diffusion model designed for real-time end-to-end autonomous driving.
It replaces the slow, multi-step denoising process of conventional diffusion policies with a lightweight anchored and truncated diffusion process, achieving over 10× faster inference while maintaining multi-modal driving behavior and state-of-the-art accuracy.

Without any extra training tricks, DiffusionDrive reaches 88.1 PDMS on the NAVSIM benchmark using the same ResNet-34 backbone, trained via imitation learning on human driving trajectories, and runs at real-time 45 FPS.
For full architectural and algorithmic details, refer to the [paper](https://arxiv.org/pdf/2411.15139)
.
The goal of this bounty is to enable efficient DiffusionDrive inference on Tenstorrent hardware, allowing users and partners to leverage high-throughput, low-latency execution for end-to-end driving policy and trajectory prediction.

### 🎯 What Success Looks Like

A successful submission will fulfill all requirements in the following stages. Payout is made after all three stages are completed.

### Stage 1 — Bring-Up

- Implement Diffusion Drive using TTNN APIs (Python).
- Model runs on either N150 or N300 hardware with no errors.
- Produces valid outputs/ego vehicle trajectories matching the implementation [here](https://github.com/hustvl/DiffusionDrive).
- Output is verifiable (visualization + simple metric checks).
- Clear instructions for setup and running the model.

### Stage 2 — Basic Optimizations

- Use optimal sharded/interleaved memory configs and sharding strategy for conv ops in backbone/heads.
- Fuse simple ops where possible (e.g., Conv+BN+Activation using TT fused ops).
- Store intermediates in L1 where beneficial.
- Use recommended TTNN/tt-metal CNN flows.
- Use TT library of fused ops like [relu with conv](https://github.com/tenstorrent/tt-metal/pull/29236)

### Stage 3 — Deeper Optimization

- Maximize core counts used per inference.
- Implement deeper TT-specific optimizations (e.g., grouped/depthwise convs, reordering to avoid unnecessary Tensor manipulations).
- Minimize memory and TM/Tensor Manipulation overheads.
- Document any advanced tuning or known issues.

### 🧭 Guidance & Starting Points

- Use the [TTNN model bring-up tech report](https://github.com/tenstorrent/tt-metal/blob/main/tech_reports/ttnn/TTNN-model-bringup.md) as your primary reference.
- for transfomer modules ViT implementation and documentation is a helpful reference: [vit documentation](https://github.com/tenstorrent/tt-metal/blob/main/tech_reports/ViT-TTNN/vit.md) [vit implmentation](https://github.com/tenstorrent/tt-metal/tree/main/models/demos/wormhole/vit)
- Reference [Yolov4 implementation (PR 29157)](https://github.com/tenstorrent/tt-metal/pull/29157) for example end-to-end flows.
- Refer to [TT fused ops](https://github.com/tenstorrent/tt-metal/pull/29236) for opportunities to optimize.
- Target input resolutions: start with 640×640×3 (or model default).
- Ask for help or file issues if ops are missing in TTNN.

### 🔎 Possible Approaches

- Start from an existing implementation of the model /the link shared above and port layers one by one to TTNN.
- Validate each submodule’s output against CPU/PyTorch reference before full integration.
- Experiment with different sharding strategies and memory configs for convolutions.
- Use TTNN profiling tools to identify bottlenecks and areas for fusion.
- Open a draft PR early to get feedback on your approach.

### 📊 Result Submission Guidelines

Beyond the model implementation itself. Contributors must submit the following material as a proof of work. However, feel free to open a PR at any time if you want us checking you are on the right track. Just understand that payout is only made after all 3 stages are completed.

- Functional model implementation
- Validation logs (output correctness)
- Performance report + header for final review

Links:
- [Link to perf sheet](https://github.com/tenstorrent/tt-metal/blob/main/tech_reports/ttnn/TTNN-model-bringup.md#41-performance-sheet)
- [Perf header](https://docs.tenstorrent.com/tt-metal/latest/ttnn/ttnn/profiling_ttnn_operations.html#perf-report-headers)

---

### 📚 Resources

- [TTNN model bring-up tech report](https://github.com/tenstorrent/tt-metal/blob/main/tech_reports/ttnn/TTNN-model-bringup.md)
- [CNN Bring-up & Optimization in TT-NN](https://github.com/tenstorrent/tt-metal/blob/main/tech_reports/CNNs/cnn_optimizations.md)
- [Yolov4 TTNN implementation PR](https://github.com/tenstorrent/tt-metal/pull/29157)
- [YOLOP docs](https://pytorch.org/hub/hustvl_yolop/)
- [TT fused ops PR](https://github.com/tenstorrent/tt-metal/pull/29236)
- [Perf report header](https://docs.tenstorrent.com/tt-metal/latest/ttnn/ttnn/profiling_ttnn_operations.html#perf-report-headers)

**Comment by sawssenbensalem1990-arch** [comment]
I'm working on this issue and will submit a PR soon.

Payment address: 0x958BD67f2f6be2Dc46D0e9e0Dd6d33F52EfCA67C

ETA: 24-48 hours

**Comment by mahmudsudo** [comment]
can i take on this ?

**Comment by jberkowitzTT** [comment]
Thanks @mahmudsudo for the interest, the bounty is now yours! Please update this issue with your progress within 2 weeks.

Review our [Bounty Terms & Conditions](https://docs.tenstorrent.com/bounty_terms.html), as these terms apply. Good luck with the project, and let us know if there are any bugs/issues on our side that are stopping you from completing the bounty. 

Let me know if you have any questions or if I can help!

**Comment by jberkowitzTT** [comment]
Hi @mahmudsudo !

Any updates on this bounty? Are you still actively working on it?

**Comment by mahmudsudo** [comment]
Yes I am

**Comment by bbberns** [comment]
## Sovereign OS: High-Resolution Optimization
Documentation improvement proposed for project stability. Payout to LTC: ltc1qx2g4z3uk2gzux77k8gkyd0aduzfx30rane4gds

**Comment by jberkowitzTT** [comment]
Hi @mahmudsudo ! Excited to see your great work, do you have an idea of when a PR will be ready?

**Comment by lmee** [comment]
I read through the full thread for "[Bounty $1500] Diffusion Drive bring up using TTNN APIs" including all comments. The core problem appears to be: This bounty is for bringing up. My root-cause hypothesis is missing conversion/lowering coverage for the target op path, plus incomplete edge-case validation. My plan is to build a deterministic failing case in the current conversion/lowering pipeline, implement the missing op mapping and validation path, then add IR/kernel regression coverage with a concise perf/compat note. I can start immediately and post repro + verification notes in-thread. Best, Jerry Li

**Comment by bitdeckai** [comment]
I can take #31269. I will share an implementation plan and first PR draft shortly.

**Comment by jberkowitzTT** [comment]
Hi @ayewo - apologies on this. Of course, we greatly appreciate your enthusiasm & great work as part of the bounty program

In an attempt to spread the bounties to as many people as possible, I assigned this to another user who's PR is in review. I see that you are assigned another bounty currently, so I can assign you a new one once that is under review.

I'm always open to feedback, so feel free to message me in GitHub, Discord or over email with any thoughts you have

Thanks!

**Comment by RustamKazmi** [comment]
I m very interested. Can you assign it to me.

**Comment by jberkowitzTT** [comment]
Hi @ayewo - any updates on this? If there's no PR by the end of the month, we'll need to look to reassign. Thanks so much for your work on this & let me know how I can help

**Comment by ayewo** [comment]
Hi @jberkowitzTT 
Thanks for the ping.

I've opened a PR: https://github.com/tenstorrent/tt-metal/pull/42995

**Comment by ayewo** [comment]
Hi @jberkowitzTT 

1. The disk on my Tenstorrent VM running on Koyeb only goes up to ~500GB but the setup files (for the download of OpenScene logs, sensor blobs, as well as nuPlan maps) used by the DiffusionDrive model requires > 500GB of free space. Koyeb supports attaching volumes to a VM but the max size (at least on my account) is only 10GB. Is there a way to provision a new VM or attach a separate volume with a higher capacity?
2. The model's real-time run of 45 FPS was achieved on a GPU and I would like to run a few tests to sanity check my implementation, so was wondering: does the promo code cover starting up a VM equiped with a GPU (NVIDIA RTX 4000) on Koyeb? Based on the screen shot below, it seems not, since an hourly price of $0.50/h is shown on the Koyeb dashboard compared to the Tenstorrent N300s price of $0.00/h (due to my applied promo code).

<img width="840" height="539" alt="Image" src="https://github.com/user-attachments/assets/c9891e01-7616-434c-b35a-a9bcf3e97e6c" />

**Comment by ayewo** [comment]
@jberkowitzTT thoughts?

**Comment by jberkowitzTT** [comment]
Hi @ayewo  Thanks for reaching out about this!

The promo code only covers the Tenstorrent N300s instance, not NVIDIA GPU instances like the RTX 4000, so the $0.50/hr you’re seeing there is expected.

On the storage side, I don’t believe we currently have a documented path for attaching a much larger self-serve volume through our Koyeb setup. If the full dataset footprint is a blocker, I’m happy to check with the Koyeb team about whether there’s a custom higher-storage option. Otherwise, if a smaller subset is enough for sanity checks, that’s likely the fastest path forward.

**Comment by ayewo** [comment]
Hi @jberkowitzTT 

I eventually had to spin up a separate VM instance with adequate resources for this.

Thanks

**Comment by ayewo** [comment]
Hi @jberkowitzTT,

A status update.

This is a pretty involved bring up. To make PR review a breeze, I've been sanity checking my work. 

First step: I decided to first replicate the salient number mentioned in the [paper](https://arxiv.org/pdf/2411.15139) in my dev environment:

> Without any extra training tricks, DiffusionDrive reaches 88.1 PDMS on the NAVSIM benchmark using the same ResNet-34 backbone

The authors obtained 88.1 PDMS on an nVidia 4090 GPU and I was able to obtain a similar number (88.04 PDMS) on a CPU as can be seen on the line:
```
            Final average score of valid results: 0.8804393469384784.
```
taken from the snipped output below:
```sh
python $NAVSIM_DEVKIT_ROOT/navsim/planning/script/run_pdm_score.py \
        train_test_split=navtest \
        agent=diffusiondrive_agent \
        worker=ray_distributed \
        agent.checkpoint_path=$CKPT \
        experiment_name=diffusiondrive_agent_eval
...
(wrapped_fn pid=458698) INFO:__main__:Processing scenario 2768 / 2922 in thread_id=d890ae72-7433-4697-8833-5f9dbfa6513b, node_id=0 [repeated 23x across cluster]
Ray objects:  25%|███████████████████████████████████████▎                                                                                                                     | 1/4 [38:18<1:54:54, 2298.09s/it]
(wrapped_fn pid=458699) INFO:__main__:Processing scenario 2792 / 3400 in thread_id=bd9cc8c3-125e-474d-a646-ff8c564aa671, node_id=0 [repeated 24x across cluster]
(wrapped_fn pid=458698) INFO:__main__:Processing scenario 2781 / 2922 in thread_id=d890ae72-7433-4697-8833-5f9dbfa6513b, node_id=0 [repeated 19x across cluster]
...
(wrapped_fn pid=458698) INFO:__main__:Processing scenario 2917 / 2922 in thread_id=d890ae72-7433-4697-8833-5f9dbfa6513b, node_id=0 [repeated 20x across cluster]
Ray objects:  50%|███████████████████████████████████████████████████████████████████████████████▌                                                                               | 2/4 [40:20<33:56, 1018.08s/it]
(wrapped_fn pid=458699) INFO:__main__:Processing scenario 2943 / 3400 in thread_id=bd9cc8c3-125e-474d-a646-ff8c564aa671, node_id=0 [repeated 20x across cluster]
(wrapped_fn pid=458699) INFO:__main__:Processing scenario 2950 / 3400 in thread_id=bd9cc8c3-125e-474d-a646-ff8c564aa671, node_id=0 [repeated 14x across cluster]
...
(wrapped_fn pid=458699) INFO:__main__:Processing scenario 3068 / 3400 in thread_id=bd9cc8c3-125e-474d-a646-ff8c564aa671, node_id=0 [repeated 12x across cluster]
Ray objects:  75%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████                                        | 3/4 [42:03<10:00, 600.40s/it]
(wrapped_fn pid=458699) INFO:__main__:Processing scenario 3075 / 3400 in thread_id=bd9cc8c3-125e-474d-a646-ff8c564aa671, node_id=0 [repeated 9x across cluster]
(wrapped_fn pid=458699) INFO:__main__:Processing scenario 3082 / 3400 in thread_id=bd9cc8c3-125e-474d-a646-ff8c564aa671, node_id=0 [repeated 7x across cluster]
...
(wrapped_fn pid=458699) INFO:__main__:Processing scenario 3399 / 3400 in thread_id=bd9cc8c3-125e-474d-a646-ff8c564aa671, node_id=0 [repeated 7x across cluster]
Ray objects: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 4/4 [46:25<00:00, 696.29s/it]
2026-06-04 08:08:45,782 INFO {/root/02/DiffusionDrive/navsim/planning/script/run_pdm_score.py:153}
        Finished running evaluation.
            Number of successful scenarios: 12146.
            Number of failed scenarios: 0.
            Final average score of valid results: 0.8804393469384784.
            Results are stored in: /root/02/exp/diffusiondrive_agent_eval/2026.06.04.07.22.11/2026.06.04.08.08.45.csv.

(wrapped_fn pid=458699) INFO:__main__:Processing scenario 3400 / 3400 in thread_id=bd9cc8c3-125e-474d-a646-ff8c564aa671, node_id=0
``` 

It took a lot of hair pulling to get to this point 😀.

Now that I have a fully functional environment, next up is confirm that my TTNN implementation tracks closely to the PyTorch reference model.

**Comment by sreeshanth-soma** [comment]
Hi @jberkowitzTT, if this reopens I am interested to work on it.

**Comment by ayewo** [comment]
@jberkowitzTT 

Without notice, the Koyeb platform rebooted my instance which wiped everything on the ephemeral 500GB drive attached to my development VM.

I'm rebuilding the VM environment all over again.

**Comment by jberkowitzTT** [comment]
Hi @ayewo - thanks for flagging this, and I'm sorry to hear that happened. Have you already reached out to support@koyeb.com?

**Comment by ayewo** [comment]
Hey @jberkowitzTT 

It set me back by a few days but I'm good now, thanks for checking in.

The Diffusion Drive model integrates into the NAVSIM benchmark to produce driving trajectories.

Wiring the TTNN model into the Diffusion Drive model agent & running it is a substantial step beyond the op port. Could you be kind enough to consider raising the bounty amount due to the extra effort required?

Thanks

**Comment by Nexussyn** [comment]
## Bounty Interest — Runtime Open Federation

**Agent:** Runtime Open Federation (AI-assisted development)
**Payment address (USDC on Base):** `0xb438d36b425b504724a1c72aa0941c80cb940995`

We are evaluating this bounty. Our system specializes in AI-assisted code analysis and implementation.

**We will post a full technical proposal within 4 hours.**
Please confirm this bounty is still available.

Payment: `0xb438d36b425b504724a1c72aa0941c80cb940995` (USDC on Base)

**Comment by jberkowitzTT** [comment]
> Hey [@jberkowitzTT](https://github.com/jberkowitzTT)
> 
> It set me back by a few days but I'm good now, thanks for checking in.
> 
> The Diffusion Drive model integrates into the NAVSIM benchmark to produce driving trajectories.
> 
> Wiring the TTNN model into the Diffusion Drive model agent & running it is a substantial step beyond the op port. Could you be kind enough to consider raising the bounty amount due to the extra effort required?
> 
> Thanks

Hi @ayewo - can you confirm that you expect to complete this by tomorrow? I can work with the engineering team on the increased reward

**Comment by ayewo** [comment]
Hi @jberkowitzTT 

> can you confirm that you expect to complete this by tomorrow? 

Yes, working earnestly towards tomorrow's deadline. I have everything working now. Mostly wrestling with differences between the TTNN environment (based on Python 3.10+ dependencies) and the NAVSIM environment (based on Python 3.9 dependencies) and testing additional optimizations.

> I can work with the engineering team on the increased reward

I'd really appreciate that. Thank you!

**Comment by velonica0** [comment]
> please provide updates & complete this by 6/24. Otherwise, we'll look to reassign.

If this task is reassigned, I would like to take it on. Thank you.

**Comment by ayewo** [comment]
@velonica0 
My [PR](https://github.com/tenstorrent/tt-metal/pull/42995) is still open and hasn't even been reviewed yet by the team.

**Comment by jberkowitzTT** [comment]
Hi @ayewo - congratulations on completing this bounty, and thank you for the great work!

I'll initiate the payment process via email soon. We deeply appreciate your contributions, and we look forward to your future involvement.

**Comment by ayewo** [comment]
Hey @jberkowitzTT 
Awesome, thank you! I learned a few things working on this.

BTW, what was the team’s response regarding my humble request earlier to which you [responded](https://github.com/tenstorrent/tt-metal/issues/31269#issuecomment-4782992486) thus:

> I can work with the engineering team on the increased reward

I’m curious.

**Comment by jberkowitzTT** [comment]
Hi @ayewo - thank you for following up, we can bump this to $2000 since it's considered a hard bounty. Thank you very much for following up on this.
